lib/ansible/vars/manager.py

```diff
@@ -661,3 +674,20 @@ def set_host_variable(self, host, varname, value):
             self._vars_cache[host] = combine_vars(self._vars_cache[host], {varname: value})
         else:
             self._vars_cache[host][varname] = value
+
+
+class VarsWithSources(dict):
+    '''
+    Dict-like class for vars that also provides source information for each var
+    '''
+    def __init__(self, data, sources):
```

Comment:
This `__init__` signature is not dict-compatible.  I know this has come up before and ther was a workaround but I'm not remembering it right this instance (and the problem might have been easier to resolve, then).

The wrokarounds I can think of right now:

* Set sources via a method.  It's not conceptually the best as sources is really something that is set once when the object is instantiated and then should not be set again.
* Use a special name with a keyword arg that minimizes the chance of conflicts.    Also not the best as there's always the chance that someone will add that as a conflicting key somewhere down the line.  If you go this route, be aware that Python2 is less flexible than Python3:
``` python
# Works on Python3 but fails on Python2:
def mine(*args, __ansible_DEADBEEF64_source='', **kwargs):    
    print(args)    
    print(__ansible_DEADBEEF64_source)    
    print(kwargs)    
    
mine()    
mine(1, 2)    
mine(1, 2, data='test')    
mine(1, 2, data='test', __ansible_DEADBEEF64_source='from user')    
mine(data='test', __ansible_DEADBEEF64_source='from user') 
```

You have to include the special variable in kwargs and then pop it to be compatible with Python2.

Reply:
The `__init__` signature doesn't need to be dict compatible, because it will never be unknowingly used in place of a dict anywhere that a new object would be created.

Reply:
*Right now.*  But that isn't something that should be depended upon in general and especially if you are creating a subclass of a builtin type.

I did recall the non-hacky way to fix this, though.  Change the constructor to be dict-compatible and then create an alternate constructor using classmethod to do what you want.  Then use the classmethod in your code.  Looks like this:

``` python
class VarsWithSources(MutableMapping):
    def __init__(self, data):
        self.sources = "<Source unset>"
        super(self, VarsWithSources).__init__(data)

    @classmethod
    def new_vars_with_source(cls, data, sources):
        new_vars_with_source = cls(data)
        new_vars_with_source.sources = sources
        return new_vars_with_source
```

Reply:
And then where you call it you do:
```python
VarsWithSources.new_vars_with_source(all_vars, _vars_sources)
```